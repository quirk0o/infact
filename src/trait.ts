import { AttributeDefinition, AttributeType, Dict } from './types'

export type AttributesOf<T> = T extends Trait<infer A, any> ? A : never
export type OptionsOf<T> = T extends Trait<any, infer O> ? O : never

export class Trait<TRes extends object = Dict, TTrans extends object = Dict> {
  static create<TRes extends object = Dict, TTrans extends object = Dict>() {
    return new Trait<TRes, TTrans>()
  }

  static compose<TRes extends object = Dict, TTrans extends object = Dict>(
    ...traits: Trait<Partial<TRes>, Partial<TTrans>>[]
  ) {
    return traits.reduce((composedTrait, trait) => composedTrait.compose(trait))
  }

  constructor(public attributes: AttributeDefinition<TRes, TTrans>[] = []) {}

  compose<F extends Trait<any, any>>(
    trait: F
  ): Trait<TRes & AttributesOf<F>, TTrans & OptionsOf<F>> {
    type AF = AttributesOf<F>
    type AO = OptionsOf<F>
    type AR = TRes & AttributesOf<F>
    type OR = TTrans & OptionsOf<F>
    return new Trait<AR, OR>(
      (this.attributes as (
        | AttributeDefinition<TRes, TTrans>
        | AttributeDefinition<AF, AO>
      )[]).concat(trait.attributes)
    )
  }

  seq<K extends keyof TRes>(
    key: K,
    initial: number = 0,
    step: number = 1
  ): (definition: (n: number, attrs: TRes & TTrans) => TRes[K]) => Trait<TRes, TTrans> {
    return definition => {
      return new Trait(
        this.attributes.concat({
          type: AttributeType.Sequence,
          key,
          step,
          get: definition,
          seq: initial
        })
      )
    }
  }

  opt<K extends keyof TTrans>(
    key: K
  ): (definition: (attrs: TRes & TTrans) => TTrans[K]) => Trait<TRes, TTrans> {
    return definition => {
      return new Trait(
        this.attributes.concat({ type: AttributeType.Transient, key, get: definition })
      )
    }
  }

  attr<K extends keyof TRes>(
    key: K
  ): (definition: (attrs: TRes & TTrans) => TRes[K]) => Trait<TRes, TTrans> {
    return definition => {
      return new Trait(
        this.attributes.concat({ type: AttributeType.Property, key, get: definition })
      )
    }
  }
}
