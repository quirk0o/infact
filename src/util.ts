export type Path = (string | symbol | number)[]

export const path = <T extends object, R = any>(path: Path) => (obj?: T): R =>
  (path.reduce((obj, key) => {
    if (!obj) return
    return (obj as any)[key]
  }, obj) as any) as R

export const mergeAll = <T extends object = object>(...objs: object[]): T =>
  objs.reduce((result, obj) => Object.assign(result, obj), {}) as T

export const isObject = <T extends object>(maybeObj?: any): maybeObj is T =>
  typeof maybeObj === 'object' && maybeObj !== null

export const concat = <T, U, M>(a1: T[], a2: U[]): M[] =>
  ((a1 as (T | U)[]).concat(a2) as any) as M[]
export const concatAll = <T, U, M>(...as: (T | U)[][]): M[] =>
  (as.reduce((acc, a) => acc.concat(a)) as any) as M[]
