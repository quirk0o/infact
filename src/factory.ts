import { lazyPropertyDescriptor, propertyDescriptor } from './object'
import { generator, Generator, GeneratorResult, take } from './generator'
import { mergeAll, path } from './util'

type Dict = { [key: string]: any }

type AfterCallback<TRes, TTrans, R extends TRes = TRes> = (
  entity: TRes,
  evaluator: TRes & TTrans
) => R

enum AttributeType {
  Sequence = 'SEQ',
  Property = 'PROP',
  Transient = 'TRANS'
}

type AttributeDefinition<TRes = Dict, TTrans = Dict> =
  | PropertyDefinition<TRes, TTrans>
  | SequenceDefinition<TRes, TTrans>
  | TransientDefinition<TRes, TTrans>

type PropertyDefinition<TRes = Dict, TTrans = Dict> = {
  type: AttributeType.Property
  key: keyof TRes
  get: (attrs: TRes & TTrans) => TRes[keyof TRes]
}
type SequenceDefinition<TRes = Dict, TTrans = Dict> = {
  type: AttributeType.Sequence
  key: keyof TRes
  get: (n: number, attrs: TRes & TTrans) => TRes[keyof TRes]
  seq: number
}
type TransientDefinition<TRes = Dict, TTrans = Dict> = {
  type: AttributeType.Transient
  key: keyof TTrans
  get: (attrs: TRes & TTrans) => TTrans[keyof TTrans]
}

type AttributesOf<F> = F extends Factory<infer A, any> ? A : never
type OptionsOf<F> = F extends Factory<any, infer O> ? O : never

const isSequence = <TRes, TTrans>(
  attribute: AttributeDefinition<TRes, TTrans>
): attribute is SequenceDefinition<TRes, TTrans> => attribute.type === AttributeType.Sequence
const isProperty = <TRes, TTrans>(
  attribute: AttributeDefinition<TRes, TTrans>
): attribute is PropertyDefinition<TRes, TTrans> => attribute.type === AttributeType.Property
const isTransient = <TRes, TTrans>(
  attribute: AttributeDefinition<TRes, TTrans>
): attribute is TransientDefinition<TRes, TTrans> => attribute.type === AttributeType.Transient

const createDescriptor = <TRes extends object = Dict, TTrans extends object = Dict>(
  key: keyof (TRes & TTrans),
  get: () => any,
  overrides?: TRes & TTrans
) =>
  path([key])(overrides)
    ? propertyDescriptor(key, overrides![key])
    : lazyPropertyDescriptor(key, get)

const sequenceDefs = <F extends Factory>(attributes: AttributesOf<F>) =>
  attributes.filter(isSequence)
const propertyDefs = <F extends Factory>(attributes: AttributesOf<F>) =>
  attributes.filter(isProperty)
const transientDefs = <F extends Factory>(attributes: AttributesOf<F>) =>
  attributes.filter(isTransient)

const descriptors = <TRes extends object = Dict, TTrans extends object = Dict>(
  attributes: (PropertyDefinition<TRes, TTrans> | TransientDefinition<TRes, TTrans>)[],
  evaluator: () => TRes & TTrans,
  overrides?: TRes & TTrans
) =>
  attributes.map(({ key, get }) =>
    createDescriptor<TRes, TTrans>(key, () => get(evaluator()), overrides)
  )
const sequenceDescriptors = <TRes extends object = Dict, TTrans extends object = Dict>(
  attributes: SequenceDefinition<TRes, TTrans>[],
  evaluator: () => TRes & TTrans,
  overrides?: TRes & TTrans
) =>
  attributes.map(({ key, get, seq }) =>
    createDescriptor<TRes, TTrans>(key, () => get(seq, evaluator()), overrides)
  )

export class Factory<TRes extends object = Dict, TTrans extends object = Dict> {
  static AttributeType = AttributeType

  static create<TRes extends object = Dict, TTrans extends object = Dict>() {
    return new Factory<TRes, TTrans>()
  }

  static compose<TRes extends object = Dict, TTrans extends object = Dict>(
    ...factories: Factory<Partial<TRes>, Partial<TTrans>>[]
  ) {
    return factories.reduce((composedFactory, factory) => composedFactory.compose(factory))
  }

  constructor(
    private attributes: AttributeDefinition<TRes, TTrans>[] = [],
    private callbacks: AfterCallback<TRes, TTrans>[] = []
  ) {}

  compose<F extends Factory<any, any>>(
    factory: F
  ): Factory<TRes & AttributesOf<F>, TTrans & OptionsOf<F>> {
    type AF = AttributesOf<F>
    type AO = OptionsOf<F>
    type AR = TRes & AttributesOf<F>
    type OR = TTrans & OptionsOf<F>
    return new Factory<AR, OR>(
      (this.attributes as (
        | AttributeDefinition<TRes, TTrans>
        | AttributeDefinition<AF, AO>
      )[]).concat(factory.attributes),
      (this.callbacks as (AfterCallback<TRes, TTrans> | AfterCallback<AF, AO>)[]).concat(
        factory.callbacks
      ) as AfterCallback<AR, OR>[]
    )
  }

  buildOne(overrides?: TRes & TTrans): TRes {
    return this.gen(overrides).next().value
  }

  gen(overrides?: TRes & TTrans): Generator<TRes, TRes & TTrans> {
    return generator<TRes, TRes & TTrans>(nextOverrides =>
      this.doGen(this.attributes, this.callbacks, nextOverrides || overrides)
    )
  }

  private doGen(
    attributes: AttributeDefinition<TRes, TTrans>[] = [],
    callbacks: AfterCallback<TRes, TTrans>[] = [],
    overrides?: TRes & TTrans
  ): GeneratorResult<TRes, TRes & TTrans> {
    let evaluator: TRes & TTrans

    const options = descriptors(transientDefs(attributes), () => evaluator, overrides)
    const sequences = sequenceDescriptors(sequenceDefs(attributes), () => evaluator, overrides)
    const properties = descriptors(propertyDefs(attributes), () => evaluator, overrides)

    const entityDescriptorMap = Object.assign({}, mergeAll(...properties), mergeAll(...sequences))
    const evaluatorDescriptorMap = Object.assign({}, entityDescriptorMap, mergeAll(...options))
    evaluator = Object.create(null, evaluatorDescriptorMap)

    const entity = Object.create(null, entityDescriptorMap)
    const modifiedEntity = callbacks.reduce(
      (newEntity, callback) => callback(newEntity, evaluator),
      entity
    )

    return {
      value: modifiedEntity,
      next: nextOverrides =>
        this.doGen(
          attributes.map(attr =>
            isSequence(attr)
              ? {
                  ...attr,
                  seq: attr.seq + 1
                }
              : attr
          ),
          callbacks,
          nextOverrides || overrides
        )
    }
  }

  buildList(n: number, overrides?: TRes & TTrans): TRes[] {
    return take(n)(this.gen(overrides))
  }

  seq<K extends keyof TRes>(
    key: K
  ): (definition: (n: number, attrs: TRes & TTrans) => TRes[K]) => Factory<TRes, TTrans> {
    return definition => {
      return new Factory(
        this.attributes.concat({ type: AttributeType.Sequence, key, get: definition, seq: 0 }),
        this.callbacks
      )
    }
  }

  opt<K extends keyof TTrans>(
    key: K
  ): (definition: (attrs: TRes & TTrans) => TTrans[K]) => Factory<TRes, TTrans> {
    return definition => {
      return new Factory(
        this.attributes.concat({ type: AttributeType.Transient, key, get: definition }),
        this.callbacks
      )
    }
  }

  attr<K extends keyof TRes>(
    key: K
  ): (definition: (attrs: TRes & TTrans) => TRes[K]) => Factory<TRes, TTrans> {
    return definition => {
      return new Factory(
        this.attributes.concat({ type: AttributeType.Property, key, get: definition }),
        this.callbacks
      )
    }
  }

  trait(): this {
    return this
  }

  after(callback: AfterCallback<TRes, TTrans>): Factory<TRes, TTrans> {
    return new Factory(this.attributes, this.callbacks.concat(callback))
  }
}
