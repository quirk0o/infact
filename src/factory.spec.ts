import { Factory } from './factory'
import { Trait } from './trait'
import * as util from 'util'

expect.addSnapshotSerializer({
  test: val => val,
  print: (val: any) => util.format(val)
})

describe('Factory', () => {
  describe('#create', () => {
    it('creates a new factory', () => {
      expect(Factory.create()).toBeInstanceOf(Factory)
    })
  })

  describe('#compose', () => {
    it('combines a set of factories', () => {
      const CatFactory = new Factory().attr('name')(() => 'Bibi')
      const SpiderFactory = new Factory().attr('canItSwing')(() => true)
      const SpiderCatFactory = Factory.compose(CatFactory, SpiderFactory)

      const spiderCat = SpiderCatFactory.build()

      expect(spiderCat).toEqual({ name: 'Bibi', canItSwing: true })
    })
  })

  describe('.build', () => {
    it('builds object with set attributes', () => {
      const CatFactory = new Factory()
        .attr('name')(() => 'Bibi')
        .attr('age')(() => 3)

      const cat = CatFactory.build()

      expect(cat).toEqual({
        name: 'Bibi',
        age: 3
      })
    })

    it('builds object with attribute overrides', () => {
      const CatFactory = new Factory()
        .attr('name')(() => 'Bibi')
        .attr('age')(() => 3)

      const cat = CatFactory.build({ age: 4 })

      expect(cat).toEqual({
        name: 'Bibi',
        age: 4
      })
    })

    it('passes options to attribute definition', () => {
      const CatFactory = new Factory()
        .opt('birthday')(() => new Date(2017, 7, 13))
        .attr('name')(() => 'Bibi')
        .attr('age')(
        ({ birthday }) =>
          (new Date(2019, 7, 13).getTime() - birthday.getTime()) / 1000 / 60 / 60 / 24 / 365
      )

      const cat = CatFactory.build()

      expect(cat).toEqual({
        name: 'Bibi',
        age: 2
      })
    })

    it('passes attributes to attribute definition', () => {
      const CatFactory = new Factory()
        .attr('name')(() => 'Bibi')
        .attr('fullName')(({ name }) => `${name} The Cat`)

      const cat = CatFactory.build()

      expect(cat).toEqual({
        name: 'Bibi',
        fullName: 'Bibi The Cat'
      })
    })

    it('passes sequences to attribute definition', () => {
      const CatFactory = new Factory()
        .seq('name')(n => `Cat #${n + 1}`)
        .attr('fullName')(({ name }) => `${name} Potato`)

      const cat = CatFactory.build()

      expect(cat).toEqual({
        name: 'Cat #1',
        fullName: 'Cat #1 Potato'
      })
    })

    it('uses option overrides to build object', () => {
      const CatFactory = new Factory()
        .opt('birthday')(() => new Date(2017, 7, 13))
        .attr('name')(() => 'Bibi')
        .attr('age')(
        ({ birthday }) =>
          (new Date(2019, 7, 13).getTime() - birthday.getTime()) / 1000 / 60 / 60 / 24 / 365
      )

      const cat = CatFactory.build({ birthday: new Date(2016, 7, 13) })

      expect(cat).toEqual({
        name: 'Bibi',
        age: 3
      })
    })

    it('memoizes attribute value', () => {
      const nameGenerator = jest
        .fn()
        .mockReturnValueOnce('Bibi')
        .mockReturnValueOnce('Fluffy')

      const CatFactory = new Factory()
        .attr('name')(nameGenerator)
        .attr('fullName')(({ name }) => `Fluffy ${name}`)

      expect(CatFactory.build()).toEqual({
        name: 'Bibi',
        fullName: 'Fluffy Bibi'
      })
    })

    it('returns a plain JS object', () => {
      const CatFactory = new Factory()
        .attr('name')(() => 'Bibi')
        .attr('fullName')(({ name }) => `Fluffy ${name}`)

      expect(CatFactory.build()).toMatchSnapshot()
    })
  })

  describe('.buildList', () => {
    it('builds a list of entities', () => {
      const CatFactory = new Factory<{ name: string; sound: string }>()
        .seq('name')(n => `Cat #${n + 1}`)
        .attr('sound')(() => 'meow')

      const cats = CatFactory.buildList(3)

      expect(cats).toEqual([
        { name: 'Cat #1', sound: 'meow' },
        { name: 'Cat #2', sound: 'meow' },
        { name: 'Cat #3', sound: 'meow' }
      ])
    })
  })

  describe('.seq', () => {
    it('increments sequence on every build', () => {
      const CatFactory = new Factory().seq('name')(n => `Cat #${n + 1}`)

      const cat = CatFactory.gen().next().value
      const anotherCat = CatFactory.gen()
        .next()
        .next().value

      expect(cat).toEqual({ name: 'Cat #1' })
      expect(anotherCat).toEqual({ name: 'Cat #2' })
    })

    describe('when initial value and step provided', () => {
      it('increments sequence by step on every build', () => {
        const CatFactory = new Factory().seq('name', 1, 2)(n => `Cat #${n}`)

        const cat = CatFactory.gen().next().value
        const anotherCat = CatFactory.gen()
          .next()
          .next().value

        expect(cat).toEqual({ name: 'Cat #1' })
        expect(anotherCat).toEqual({ name: 'Cat #3' })
      })
    })
  })

  describe('.after', () => {
    describe('after build', () => {
      it('invokes after callbacks and returns modified object', () => {
        const afterCallback = jest.fn().mockImplementation((cat, evaluator) =>
          evaluator.hungry
            ? {
                ...cat,
                meowing: true
              }
            : cat
        )
        const CatFactory = new Factory()
          .attr('name')(() => 'Bibi')
          .opt('hungry')(() => true)
          .after(afterCallback)

        const cat = CatFactory.build()

        expect(afterCallback).toHaveBeenCalled()
        expect(cat).toEqual({ name: 'Bibi', meowing: true })
      })

      it('invokes after callbacks in order they are added', () => {
        const afterCallbackMeowing = jest.fn().mockImplementation((cat, evaluator) =>
          evaluator.hungry
            ? {
                ...cat,
                meowing: true
              }
            : cat
        )
        const afterCallbackName = jest.fn().mockImplementation(cat =>
          cat.meowing
            ? {
                ...cat,
                name: `Meowing ${cat.name}`
              }
            : cat
        )
        const CatFactory = new Factory()
          .attr('name')(() => 'Bibi')
          .opt('hungry')(() => true)
          .after(afterCallbackMeowing)
          .after(afterCallbackName)

        const cat = CatFactory.build()

        expect(afterCallbackMeowing).toHaveBeenCalled()
        expect(afterCallbackName).toHaveBeenCalled()
        expect(cat).toEqual({ name: 'Meowing Bibi', meowing: true })
      })

      it('returns unmodified object when nothing is returned from callback', () => {
        const callback = jest.fn()
        const CatFactory = new Factory()
          .attr('name')(() => 'Bibi')
          .after(callback)

        const cat = CatFactory.build()
        expect(callback).toHaveBeenCalled()
        expect(cat).toEqual({ name: 'Bibi' })
      })
    })
  })

  describe('.trait', () => {
    it('supports all attribute methods of Factory', () => {
      const CatFactory = new Factory().trait('super')(t =>
        t
          .seq(
            'snacksEaten',
            1,
            2
          )(n => n)
          .attr('name')(({ hungry }) => (hungry ? 'Hungry Bibi' : 'Super Bibi'))
          .opt('hungry')(() => false)
      )

      const cat = CatFactory.build('super', { hungry: true })

      expect(cat).toEqual({
        snacksEaten: 1,
        name: 'Hungry Bibi'
      })
    })

    it('overrides previously declared attributes', () => {
      const CatFactory = new Factory()
        .attr('sound')(() => 'meow')
        .trait('mad')(t => t.attr('sound')(() => 'raaawrr'))

      expect(CatFactory.build()).toEqual({ sound: 'meow' })
      expect(CatFactory.build('mad')).toEqual({ sound: 'raaawrr' })
    })

    describe('when passed a function', () => {
      it('creates a new trait', () => {
        const CatFactory = new Factory()
          .attr('name')(() => 'Bibi')
          .attr('age')(() => 4)
          .trait('super')(t =>
          t
            .attr('name')(() => 'Super Bibi')
            .attr('power')(() => 'High Pitched Meow')
        )

        const cat = CatFactory.build('super', { age: 3 })

        expect(cat).toEqual({
          name: 'Super Bibi',
          age: 3,
          power: 'High Pitched Meow'
        })
      })
    })

    describe('when passed a trait', () => {
      it('uses the trait', () => {
        const CatFactory = new Factory()
          .attr('name')(() => 'Bibi')
          .attr('age')(() => 4)
          .trait('super')(
          new Trait()
            .attr('name')(() => 'Super Bibi')
            .attr('power')(() => 'High Pitched Meow')
        )

        const cat = CatFactory.build('super', { age: 3 })

        expect(cat).toEqual({
          name: 'Super Bibi',
          age: 3,
          power: 'High Pitched Meow'
        })
      })
    })

    describe('when passed a mixed trait', () => {
      it('uses the trait', () => {
        const SuperTrait = new Trait().attr('power')(() => 'Superpower')
        const SpiderTrait = new Trait().attr('canSwing')(() => true)

        const SuperSpiderTrait = Trait.compose(SuperTrait, SpiderTrait)

        const SpiderCatFactory = new Factory()
          .attr('name')(() => 'Bibi')
          .attr('age')(() => 4)
          .trait('super')(SuperSpiderTrait)

        const cat = SpiderCatFactory.build('super', { age: 3 })

        expect(cat).toEqual({
          name: 'Bibi',
          age: 3,
          power: 'Superpower',
          canSwing: true
        })
      })
    })
  })
})
