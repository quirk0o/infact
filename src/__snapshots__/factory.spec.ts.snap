// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Factory .build returns a plain JS object 1`] = `{ name: 'Bibi', fullName: 'Fluffy Bibi' }`;
